p2pnetwork/nodeconnection.py

```diff
@@ -187,38 +188,28 @@ def run(self):
         """The main loop of the thread to handle the connection with the node. Within the
            main loop the thread waits to receive data from the node. If data is received 
            the method node_message will be invoked of the main node to be processed."""
-        buffer = b''  # Hold the stream that comes in!
-
         while not self.terminate_flag.is_set():
-            chunk = b''
-
             try:
-                chunk = self.sock.recv(4096)
+                packet_length = self.sock.recv(8)
+                if len(packet_length) > 0:
+                    (unpacked_length,) = unpack('>Q', packet_length)
+                    buffer = b''
+                    # while the length of the buffer is smaller than the length of the package
+                    while len(buffer) < unpacked_length:
+                        to_read = unpacked_length - len(buffer)
+                        buffer += self.sock.recv(4096 if to_read > 4096 else to_read)
+                        time.sleep(0.01)
 
+                    self.main_node.message_count_recv += 1
+                    self.main_node.node_message(self, self.parse_packet(buffer))
+                time.sleep(0.01)
             except socket.timeout:
```

Comment:
At this moment, here it will get when this occurs. However, we should never relay on exceptions for our "happy flow" or main code. Anyway, it will be not causing big hang ups ;).

Reply:
Agreed